backends/ent/annotations.go

```diff
@@ -0,0 +1,215 @@
+	tx, err := backend.txClient()
+	if err != nil {
+		return err
+	}
+
+	builders := []*ent.AnnotationCreate{}
+
+	for idx := range data {
+		builder := tx.Annotation.Create().
+			SetDocumentID(data[idx].DocumentID).
+			SetName(data[idx].Name).
+			SetValue(data[idx].Value)
+
+		builders = append(builders, builder)
+	}
+
+	err = tx.Annotation.CreateBulk(builders...).
+		OnConflict().
+		UpdateNewValues().
+		Exec(backend.ctx)
+	if err != nil && !ent.IsConstraintError(err) {
+		return rollback(tx, fmt.Errorf("creating annotations: %w", err))
+	}
+
+	if err := tx.Commit(); err != nil {
+		return rollback(tx, err)
+	}
+
+	return nil
+}
+
+// AddAnnotations applies multiple named annotation values to a single document.
+func (backend *Backend) AddAnnotations(documentID, name string, values ...string) error {
+	data := ent.Annotations{}
+	for _, value := range values {
+		data = append(data, &ent.Annotation{
+			DocumentID: documentID,
+			Name:       name,
+			Value:      value,
+		})
+	}
+
+	return backend.createAnnotations(data...)
+}
+
+// AddAnnotationToDocuments applies a single named annotation value to multiple documents.
+func (backend *Backend) AddAnnotationToDocuments(name, value string, documentIDs ...string) error {
+	data := ent.Annotations{}
+	for _, documentID := range documentIDs {
+		data = append(data, &ent.Annotation{
+			DocumentID: documentID,
+			Name:       name,
+			Value:      value,
+		})
+	}
+
+	return backend.createAnnotations(data...)
+}
+
+// ClearAnnotations removes all annotations from the specified documents.
+func (backend *Backend) ClearAnnotations(documentIDs ...string) error {
+	tx, err := backend.txClient()
+	if err != nil {
+		return err
+	}
+
+	_, err = tx.Annotation.Delete().
+		Where(annotation.DocumentIDIn(documentIDs...)).
+		Exec(backend.ctx)
+	if err != nil {
+		return rollback(tx, fmt.Errorf("clearing annotations: %w", err))
+	}
+
+	if err := tx.Commit(); err != nil {
+		return rollback(tx, err)
+	}
+
+	return nil
+}
+
+// GetDocumentAlias gets the value for the annotation named "alias".
+func (backend *Backend) GetDocumentAlias(documentID string) (string, error) {
+	if backend.client == nil {
+		return "", errUninitializedClient
+	}
+
+	query := backend.client.Annotation.Query().
+		Where(
+			annotation.And(
+				annotation.HasDocumentWith(document.IDEQ(documentID)),
+				annotation.NameEQ("alias"),
+			),
+		)
+
+	result, err := query.Only(backend.ctx)
+	if err != nil {
+		if ent.IsNotFound(err) {
+			return "", nil
+		}
+
+		return "", fmt.Errorf("retrieving document alias: %w", err)
+	}
+
+	return result.Value, nil
+}
+
+// GetDocumentAnnotations gets all annotations for the specified
+// document, limited to a set of annotation names if specified.
+func (backend *Backend) GetDocumentAnnotations(documentID string, names ...string) (ent.Annotations, error) {
+	if backend.client == nil {
+		return nil, errUninitializedClient
+	}
```

Comment:
I've noticed the transaction client get initialized on the fly while this one errs when undefined. Any reason for the inconsistency?

Reply:
The transactional client is currently only used for batching create/update/delete operations so it can be safely rolled back before committing any changes. Read operations just use the regular client since nothing is being altered. The `txClient` function also starts off by running this check, so it seemed consistent to me